pkg/core/blockchain.go

```diff
@@ -613,18 +623,225 @@ func (bc *Blockchain) jumpToStateInternal(p uint32, stage stateJumpStage) error
 		Root:  block.PrevStateRoot,
 	})
 
+	bc.dao.Store.Delete(jumpStageKey)
+
+	err = bc.resetRAMState(p, false)
+	if err != nil {
+		return fmt.Errorf("failed to update in-memory blockchain data: %w", err)
+	}
+	return nil
+}
+
+// resetRAMState resets in-memory cached info.
+func (bc *Blockchain) resetRAMState(height uint32, resetHeaders bool) error {
+	if resetHeaders {
+		bc.headerHashes = bc.headerHashes[:height+1]
+		bc.storedHeaderCount = height + 1
+	}
+	block, err := bc.dao.GetBlock(bc.headerHashes[height])
+	if err != nil {
+		return fmt.Errorf("failed to get current block: %w", err)
+	}
+	bc.topBlock.Store(block)
+	atomic.StoreUint32(&bc.blockHeight, height)
+	atomic.StoreUint32(&bc.persistedHeight, height)
+
 	err = bc.initializeNativeCache(block.Index, bc.dao)
 	if err != nil {
 		return fmt.Errorf("failed to initialize natives cache: %w", err)
 	}
 
-	if err := bc.updateExtensibleWhitelist(p); err != nil {
+	if err := bc.updateExtensibleWhitelist(height); err != nil {
 		return fmt.Errorf("failed to update extensible whitelist: %w", err)
 	}
 
-	updateBlockHeightMetric(p)
+	updateBlockHeightMetric(height)
+	return nil
+}
 
-	bc.dao.Store.Delete(jumpStageKey)
+// Reset resets chain state to the specified height if possible. This method
+// performs rough DB changes and can be called on non-running Blockchain only.
```

Comment:
`direct DB changes`? BTW, can we check if the chain is active and return an error if so?